File: frontend/src/types/workflow.ts
export interface Node {
    id: string;
    type: string;        // trigger, action, agent, tool, data, logic, human
    sub_type: string;    // webhook, schedule, agent_call, http_request, condition, etc.
    label: string | null;
    config_json: string;
    position_x: string;
    position_y: string;
    workflow_id?: string;
}

export interface Edge {
    id: string;
    source_node_id: string;
    target_node_id: string;
    source_handle?: string | null;
    label?: string | null;
    workflow_id?: string;
}

export interface Workflow {
    id: string;
    name: string;
    description: string | null;
    is_active: boolean;
    agent_id: string | null;
    channel_id: string | null;
    version: string | null;
    created_at: string;
    updated_at: string | null;
}

export interface WorkflowGraph extends Workflow {
    nodes: Node[];
    edges: Edge[];
}

// ─── Execution Types ─────────────────────────────────────

export interface WorkflowRun {
    id: string;
    workflow_id: string;
    status: 'pending' | 'running' | 'completed' | 'failed' | 'paused';
    trigger_payload: string;
    error: string | null;
    started_at: string | null;
    ended_at: string | null;
}

export interface NodeExecution {
    id: string;
    run_id: string;
    node_id: string;
    status: 'waiting' | 'running' | 'completed' | 'failed' | 'skipped';
    input_json: string;
    output_json: string;
    error: string | null;
    started_at: string | null;
    ended_at: string | null;
}

export interface WorkflowRunDetail extends WorkflowRun {
    node_executions: NodeExecution[];
}

// ─── Node Category Colors ────────────────────────────────

export const NODE_CATEGORY_COLORS: Record<string, string> = {
    trigger: '#22c55e',    // Green
    agent: '#a855f7',      // Purple
    tool: '#3b82f6',       // Blue
    data: '#06b6d4',       // Cyan
    logic: '#eab308',      // Yellow
    human: '#f97316',      // Orange
    action: '#6366f1',     // Indigo (legacy)
};

export const NODE_CATEGORY_BG: Record<string, string> = {
    trigger: 'rgba(34, 197, 94, 0.12)',
    agent: 'rgba(168, 85, 247, 0.12)',
    tool: 'rgba(59, 130, 246, 0.12)',
    data: 'rgba(6, 182, 212, 0.12)',
    logic: 'rgba(234, 179, 8, 0.12)',
    human: 'rgba(249, 115, 22, 0.12)',
    action: 'rgba(99, 102, 241, 0.12)',
};

File: backend/app/models/workflow.py
from sqlalchemy import Column, String, Boolean, DateTime, ForeignKey, Text, Index
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import uuid
from app.db.base import Base

def generate_uuid():
    return str(uuid.uuid4())


class Workflow(Base):
    __tablename__ = "workflows"

    id = Column(String, primary_key=True, default=generate_uuid)
    name = Column(String, nullable=False)
    description = Column(String, nullable=True)
    is_active = Column(Boolean, default=True)
    agent_id = Column(String, ForeignKey("agents.id", ondelete="SET NULL"), nullable=True)
    channel_id = Column(String, ForeignKey("channels.id", ondelete="SET NULL"), nullable=True)
    version = Column(String, default="1")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    nodes = relationship("Node", back_populates="workflow", cascade="all, delete-orphan")
    edges = relationship("Edge", back_populates="workflow", cascade="all, delete-orphan")
    runs = relationship("WorkflowRun", back_populates="workflow", cascade="all, delete-orphan")


class Node(Base):
    __tablename__ = "nodes"

    id = Column(String, primary_key=True)  # Using ReactFlow string IDs
    workflow_id = Column(String, ForeignKey("workflows.id", ondelete="CASCADE"), nullable=False)
    type = Column(String, nullable=False)        # trigger, action, agent, tool, data, logic, human
    sub_type = Column(String, nullable=False)     # webhook, schedule, agent_call, http_request, etc.
    label = Column(String, nullable=True)         # Human-readable display name
    config_json = Column(Text, default="{}")      # JSON string for configuration

    # UI Positioning
    position_x = Column(String, default="0")
    position_y = Column(String, default="0")

    workflow = relationship("Workflow", back_populates="nodes")


class Edge(Base):
    __tablename__ = "edges"

    id = Column(String, primary_key=True) # ReactFlow edge ID
    workflow_id = Column(String, ForeignKey("workflows.id", ondelete="CASCADE"), nullable=False)
    source_node_id = Column(String, nullable=False)
    target_node_id = Column(String, nullable=False)
    source_handle = Column(String, nullable=True)   # For multi-output nodes (e.g. condition true/false)
    label = Column(String, nullable=True)            # Optional edge label

    workflow = relationship("Workflow", back_populates="edges")


class WorkflowRun(Base):
    """Tracks each execution of a workflow."""
    __tablename__ = "workflow_runs"
    __table_args__ = (
        Index("ix_workflow_runs_status", "status"),
        Index("ix_workflow_runs_workflow_id", "workflow_id"),
    )

    id = Column(String, primary_key=True, default=generate_uuid)
    workflow_id = Column(String, ForeignKey("workflows.id", ondelete="CASCADE"), nullable=False)
    status = Column(String, default="pending")       # pending, running, completed, failed, paused
    trigger_payload = Column(Text, default="{}")     # JSON of the triggering data
    context_json = Column(Text, default="{}")        # JSON of the runtime execution context

    # Checkpoint support for resuming paused runs
    checkpoint_node_id = Column(String, nullable=True)
    checkpoint_payload = Column(Text, nullable=True)

    error = Column(Text, nullable=True)
    started_at = Column(DateTime(timezone=True), server_default=func.now())
    ended_at = Column(DateTime(timezone=True), nullable=True)

    workflow = relationship("Workflow", back_populates="runs")
    node_executions = relationship("NodeExecution", back_populates="run", cascade="all, delete-orphan")


class NodeExecution(Base):
    """Tracks execution of each node within a workflow run."""
    __tablename__ = "node_executions"

    id = Column(String, primary_key=True, default=generate_uuid)
    run_id = Column(String, ForeignKey("workflow_runs.id", ondelete="CASCADE"), nullable=False)
    node_id = Column(String, nullable=False)          # References Node.id
    status = Column(String, default="waiting")        # waiting, running, completed, failed, skipped
    input_json = Column(Text, default="{}")
    output_json = Column(Text, default="{}")
    error = Column(Text, nullable=True)
    started_at = Column(DateTime(timezone=True), nullable=True)
    ended_at = Column(DateTime(timezone=True), nullable=True)

    run = relationship("WorkflowRun", back_populates="node_executions")


class WorkflowMemory(Base):
    """Persistent memory associated with a workflow, partitioned by agent or channel."""
    __tablename__ = "workflow_memories"

    id = Column(String, primary_key=True, default=generate_uuid)
    workflow_id = Column(String, ForeignKey("workflows.id", ondelete="CASCADE"), nullable=False)
    agent_id = Column(String, ForeignKey("agents.id", ondelete="CASCADE"), nullable=True)
    channel_id = Column(String, ForeignKey("channels.id", ondelete="CASCADE"), nullable=True)

    memory_json = Column(Text, default="{}")
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())
